### ISSUE TITLE
[BUG] fix math behind update_change_target_from_false in mask.py

### ISSUE THREAD

**Issue by sritchie** [genjax-community/genjax#1230]
Here was the broken, old implementation:

```python
    @typecheck
    def update_change_target_from_false(
        self,
        key: PRNGKey,
        trace: Trace,
        update_problem: UpdateProblem,
        argdiffs: Argdiffs,
    ) -> tuple[Trace, Weight, Retdiff, UpdateProblem]:
        check = Diff.tree_primal(argdiffs)[0]
        check_diff, inner_argdiffs = argdiffs[0], argdiffs[1:]

        inner_trace = EmptyTrace(self.gen_fn)

        assert isinstance(update_problem, Constraint)
        imp_update_problem = ImportanceProblem(update_problem)

        premasked_trace, w, _, _ = self.gen_fn.update(
            key, inner_trace, GenericProblem(tuple(inner_argdiffs), imp_update_problem)
        )

        _, _, retdiff, bwd_problem = self.gen_fn.update(
            key, premasked_trace, GenericProblem(tuple(inner_argdiffs), update_problem)
        )

        w = jax.lax.select(
            check,
            premasked_trace.get_score(),
            0.0,
        )

        return (
            MaskTrace(self, premasked_trace, check),
            w,
            Mask.maybe(check_diff, retdiff),
            MaskedProblem(check, bwd_problem),
        )
```

**Comment by georgematheos** [comment]
@sritchie Tomorrow we are thinking of starting to use the masking system in ChiSight.  Will this bug be blocking for us?  If so, can you recommend any workaround for now (e.g. an alternative implementation of the Mask combinator that we should add to our codebase)?

**Comment by derifatives** [comment]
@arijitnoobstar created a workaround, which you can patch in via (modify the `cp` this to point your GenJAX of course):
```
curl https://gist.githubusercontent.com/arijitnoobstar/feccb62d8914bf763073ef4ea3a83862/raw > /tmp/mask.py
cp /tmp/mask.py /opt/conda/envs/hgps/lib/python3.11/site-packages/genjax/_src/generative_functions/combinators/mask.py 
```

**Comment by MathieuHuot** [comment]
I think Arijit's fix was integrated in [this PR](https://github.com/probcomp/genjax/pull/1204) which is merged on the latest main but maybe not on the latest 0.5.1 release (if it's true @sritchie I think we should publish a 0.5.2 ASAP). Later McCoy noted there may still be a math bug in one case and that's why this issue is still open.

**Comment by georgematheos** [comment]
Oh, great!  @arijitnoobstar can you please confirm that the PR that Mathieu linked to resolves the issue?

**Comment by sritchie** [comment]
@femtomc I think we're going to need to pair on getting the correct spot from your other PR ported over...

**Comment by MathieuHuot** [comment]
> @femtomc I think we're going to need to pair on getting the correct spot from your other PR ported over...

@sritchie I thought we agreed us two could pair on this?

**Comment by femtomc** [comment]
I provide a fix for this in #1322, which passes the tests which @MathieuHuot and @arijitnoobstar added for their PR -- but we need to add new tests for this branch in particular.
